app/src/lib/git/status.ts

```diff
@@ -198,3 +158,97 @@ export async function getStatus(
     workingDirectory,
   }
 }
+
+/** reducer(ish) to create the map of file change statuses
+ * from a list of entries
+ */
+function addEntryToFiles(
+  files: Map<string, WorkingDirectoryFileChange>,
+  entry: IStatusEntry,
+  filesWithConflictMarkers: Set<string>
+) {
+  const status = mapStatus(entry.statusCode)
+  const hasConflictMarkers = filesWithConflictMarkers.has(entry.path)
+  if (status.kind === 'ordinary') {
+    // when a file is added in the index but then removed in the working
+    // directory, the file won't be part of the commit, so we can skip
+    // displaying this entry in the changes list
+    if (
+      status.index === GitStatusEntry.Added &&
+      status.workingTree === GitStatusEntry.Deleted
+    ) {
+      return files
+    }
+  }
+
+  if (status.kind === 'untracked') {
+    // when a delete has been staged, but an untracked file exists with the
+    // same path, we should ensure that we only draw one entry in the
+    // changes list - see if an entry already exists for this path and
+    // remove it if found
+    files.delete(entry.path)
+  }
+
+  // for now we just poke at the existing summary
+  const summary = convertToAppStatus(status, hasConflictMarkers)
+  const selection = DiffSelection.fromInitialSelection(DiffSelectionType.All)
+
+  files.set(
+    entry.path,
+    new WorkingDirectoryFileChange(
+      entry.path,
+      summary,
+      selection,
+      entry.oldPath
+    )
+  )
+  return files
+}
+
+/** reducer to calculate the ahead / behind and branch names
```

Comment:
Like above, we should want the docs to describe what the function does, rather than the broader environment it's being used in.

A suggested alternative:


```ts
/**
 * Update status header based on the current header entry
 */
function parseStatusHeader(
  results: StatusBuilder,
  header: IStatusHeader
) {
```

I also extracted out `StatusBuilder` (feel free to name it something else if inspiration strikes, it's not quite the Builder pattern here) as it feels important for internal consistency:

```ts
type StatusBuilder = {
  currentBranch: string | undefined
  currentUpstreamBranch: string | undefined
  currentTip: string | undefined
  branchAheadBehind: IAheadBehind | undefined
  m: RegExpMatchArray | null
}
```